### ISSUE TITLE
App freezes when paying "hold" invoices

### ISSUE THREAD

**Issue by boston-wine** [ZeusLN/zeus#1865]
### Describe the bug

Trying to pay a lightning "hold" invoice – for example, one that might be used as an escrow bond – the app freezes after clicking send, and requires restarting in order to function normally.

### Reproduce

When using Zeus to pay a hold invoice (the type that might be utilized for lightning-escrowed funds or bonds), the app froze  on "sending payment". A minute or two later, the bond invoice (on the receiving end) updated to show that it had received the funds, but Zeus was still stuck on "sending". After a few minutes with no change, I restarted the app. At this point, the payment history showed the hold invoice correctly as pending, and a while later, correctly showed the payment as expired.

I replicated this a second time, with essentially the same steps, both with relatively small (under 50K sat) invoices.

### ZEUS version

v0.8.0

### Node interface

Embedded LND

### Network

Clearnet

### Device

iPhone 11

### Device operating system

16.7.1

### Log output

```shell
how private is this?
```

**Comment by pcwd** [comment]
Hang screen looks like it's caused by this error.

ReactNativeBlobUtil failed to encode response data to BASE64 string

**Comment by AndySchroder** [comment]
> A minute or two later, the bond invoice (on the receiving end) updated to show that it had received the funds,

you mean it released the preimage or just received the HTLC?

>but Zeus was still stuck on "sending". After a few minutes with no change, I restarted the app. At this point, the payment history showed the hold invoice correctly as pending, and a while later, correctly showed the payment as expired.

So, did the receiving end cancel the HOLD invoice or accept it?

**Comment by AndySchroder** [comment]
If I do the following

1. Display HOLD invoice
2. Scan and pay hold invoice
3. Zeus says "Sending Transaction" with yellow squiggle animation.
4. Release preimage after **50** seconds
5. Get `Transaction successfully sent`

or

1. Display HOLD invoice
2. Scan and pay hold invoice
3. Zeus says "Sending Transaction" with yellow squiggle animation.
4. Release preimage after **60** seconds
5. Get `ReactNativeBlobUtil failed to encode response data to BASE64 string` error. The "Activity" history screen does show the transaction worked though.


In both cases, the app does not actually freeze.

In both cases, I would still expect a better UX.

Is there any way to know it is a HOLD invoice and allow the user to interact with it more cleanly? A hold invoice could be open for **way** longer than the 60 seconds of my test here.

I am using v0.8.1

**Comment by AndySchroder** [comment]
I did just once get a screen to come up that says "Payment is in transit. It may complete momentarily or it may go on to fail. Please check back later." I am not sure what triggers this scenario.

**Comment by AndySchroder** [comment]
Note: I also tested with the following wallets yesterday:

1. phoenix
2. breez
3. mutiny
4. muun
5. strike
6. cash app

They all had a cleaner UX with the hold invoices where the hold invoice gracefully went to a pending or processing state in their transaction list.

Once the preimage was released, they transitioned to a complete or processed state in their transaction history.

If the hold invoice was cancelled, most of them seemed to use a word like "failed", but it wasn't a big nasty screen like in Zeus (https://github.com/ZeusLN/zeus/issues/2001), it was just in the transaction history.

Generally, the wallets that seemed more intuitive had some short/recent transaction history on the home screen, so you didn't have to go to a separate Activity screen to see how the held invoices were doing.

**Comment by AndySchroder** [comment]
Would like to add a bounty of 200,000 sat for fixing this issue.

**Comment by kaloudis** [comment]
Status update on this one.

The hard freeze as originally reported no longer occurs on current builds: embedded LND payment dispatch races a forced client-side timeout, so once the payment timeout window elapses the app returns a "payment timed out, may be in transit" message instead of hanging on "sending" until restart.

What remained is that this message was a dead end. A hold invoice payment stays in flight until the recipient settles or cancels, and Zeus did not follow it to its outcome. Recent work covering that:

- #4306 (merged): fixed premature client-side timeouts on remote LND REST, CLNRest, LndHub, and Tor, so slow or held payments surface the in-transit message instead of a generic transport error.
- #4146 (merged): hold invoice payments made through the NWC service role are reported as pending instead of failed.
- #4492 (open): tracks in-flight payments to a terminal state. On embedded LND, the interface reported here, a payment still pending at the timeout now shows the in-transit screen and is polled against the node until it succeeds or fails, so a hold invoice that settles minutes later flips to success without a restart. Tracking gives up after 10 minutes, since a held HTLC can pend until CLTV expiry; the payment then remains visible as pending in the Activity view.
- #4495 (open, stacked on #4492): extends the same lookup and tracking to CLN, LDK Node, and NWC client backends.

Leaving this open until #4492 lands.

**Comment by AndySchroder** [comment]
> The hard freeze as originally reported no longer occurs on current builds: embedded LND payment dispatch races a forced client-side timeout, so once the payment timeout window elapses the app returns a "payment timed out, may be in transit" message instead of hanging on "sending" until restart.

Yes, currently getting this error with v13.2.1:

<img width="1080" height="2286" alt="Image" src="https://github.com/user-attachments/assets/28d93053-c719-4025-8c57-9dc6492e0c5f" />

**Comment by AndySchroder** [comment]
> instead of hanging on "sending" until restart

I don't think it should hang on "sending", but do something nicer like the other wallets I mention in https://github.com/ZeusLN/zeus/issues/1865#issuecomment-1964182761 do.

**Comment by kaloudis** [comment]
> > instead of hanging on "sending" until restart
> 
> I don't think it should hang on "sending", but do something nicer like the other wallets I mention in [#1865 (comment)](https://github.com/ZeusLN/zeus/issues/1865#issuecomment-1964182761) do.

How is this different than what we're doing now?

Pending payments show as an hour glass in the wallet header and show in the the activity list both as in transit and failed if those settings are enabled.

**Comment by AndySchroder** [comment]
> Pending payments show as an hour glass in the wallet header and show in the the activity list both as in transit and failed if those settings are enabled.

This part is good and similar to what the other wallets I referenced above do, but the difference is that those wallets gracefully return back to the main screen after a little bit rather than just stay on a dedicated "sending" page (potentially for hours).

So I think the logic of the other wallets is, let's try for a while and if we get no immediate error, then likely this payment is going to take a while and let's give the user some indication that they should not panic, but instead need to wait for the payment to settle.
